src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala

```diff
@@ -286,9 +290,126 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
+  }
+
+  private def insertPolicyRoles(roles: Set[ResourceRoleName], policyId: PolicyPK)(implicit session: DBSession): Int = {
+    val rr = ResourceRoleTable.syntax("rr")
+    val prCol = PolicyRoleTable.column
+    samsql"""insert into ${PolicyRoleTable.table} (${prCol.resourcePolicyId}, ${prCol.resourceRoleId})
+            select ${policyId}, ${rr.result.id} from ${ResourceRoleTable as rr} where ${rr.role} in (${roles})"""
+      .update().apply()
+  }
+
+  private def insertPolicy(policy: AccessPolicy, groupId: GroupPK)(implicit session: DBSession): PolicyPK = {
+    val pCol = PolicyTable.column
+    PolicyPK(samsql"""insert into ${PolicyTable.table} (${pCol.resourceId}, ${pCol.groupId}, ${pCol.public}, ${pCol.name})
+              values ((${loadResourcePK(policy.id.resource)}), ${groupId}, ${policy.public}, ${policy.id.accessPolicyName})""".updateAndReturnGeneratedKey().apply())
+  }
+
+  private def insertPolicyGroup(policy: AccessPolicy)(implicit session: DBSession): GroupPK = {
+    val gCol = GroupTable.column
+
+    val r = ResourceTable.syntax("r")
+    val rt = ResourceTypeTable.syntax("rt")
+    // Group.name in DB cannot be null and must be unique.  Policy names are stored in the SAM_POLICY table, and
+    // policies don't care about the group.name, but it must be set.  So we are building something unique here, in order
+    // to satisfy the db constraint, but it's otherwise irrelevant for polices.
+    val policyGroupName =
+      samsqls"""select concat(${r.id}, '_', ${policy.id.accessPolicyName}) from ${ResourceTable as r}
+               join ${ResourceTypeTable as rt} on ${r.resourceTypeId} = ${rt.id}
+               where ${r.name} = ${policy.id.resource.resourceId}
+               and ${rt.name} = ${policy.id.resource.resourceTypeName}"""
+
+    GroupPK(samsql"""insert into ${GroupTable.table} (${gCol.name}, ${gCol.email}, ${gCol.updatedDate})
+               values ((${policyGroupName}), ${policy.email}, ${Instant.now()})""".updateAndReturnGeneratedKey().apply())
+  }
+
+  private def loadResourcePK(resource: FullyQualifiedResourceId): SQLSyntax = {
+    val r = ResourceTable.syntax("r")
+    val rt = ResourceTypeTable.syntax("rt")
+    samsqls"select ${r.id} from ${ResourceTable as r} join ${ResourceTypeTable as rt} on ${r.resourceTypeId} = ${rt.id} where ${r.name} = ${resource.resourceId} and ${rt.name} = ${resource.resourceTypeName}"
+  }
+
+  override def deletePolicy(policy: FullyQualifiedPolicyId): IO[Unit] = {
+    val p = PolicyTable.syntax("p")
+    val g = GroupTable.syntax("g")
+
+    runInTransaction { implicit session =>
+      samsql"""delete from ${GroupTable as g}
+        using ${PolicyTable as p}
+        where ${g.id} = ${p.groupId}
+        and ${p.name} = ${policy.accessPolicyName}
+        and ${p.resourceId} = (${loadResourcePK(policy.resource)})""".update().apply()
+    }
+  }
+
+  override def loadPolicy(resourceAndPolicyName: FullyQualifiedPolicyId): IO[Option[AccessPolicy]] = {
```

Comment:
I don't think you should change anything for this PR.

This is awesome that you got this to work with a single query, it's probably pretty performant, but I do find it challenging to grok.  I think it would become easier if we broke it down into subqueries, maybe one for policy info, one for members, one for roles, and one for actions.  If we had the time, I'd be curious to see the perf difference to see if we could balance readability with perf.